### PR TITLE
Transaction with delete then update should not fail

### PR DESCRIPTION
**Pull request by jamesagnew**
Per this mailing list discussion: https://groups.google.com/d/msgid/hapi-fhir/8c53aa54-80e2-4616-bdb0-5bc47a50eb34n%40googlegroups.com?utm_medium=email&utm_source=footer

A transaction with a delete followed by an update on the same resource currently fails. This PR fixes that, and optimizes deletes a bit too.

### CODE REVIEW

**Review by jmarchionatto — APPROVED** [review]
Great fixes!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 1]
```diff
@@ -1091,9 +1091,17 @@ public ResourceTable updateEntity(RequestDetails theRequest, final IBaseResource
 			return entity;
 		}
 
-		if (theUpdateVersion) {
+		if (theUpdateVersion && !entity.isVersionUpdatedInCurrentTransaction()) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1094]: I think I have to admit it's dangerous for me too. I spent ages truthfully on this issue because I didn't actually understand what hibernate was doing.

I think you're right, manipulating this field is dangerous. I'm going to rework this to not require us doing that.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java [hunk 2]
```diff
@@ -349,40 +351,48 @@ private DaoMethodOutcome doCreateForPostOrPut(RequestDetails theRequest, T theRe
...
+					};
+
+					Supplier<IIdType> idSupplier = () -> {
+						return myTxTemplate.execute(tx -> {
+							IIdType retVal = myIdHelperService.translatePidIdToForcedId(myFhirContext, myResourceName, pid);
+							if (!retVal.hasVersionIdPart()) {
+								Long version = myMemoryCacheService.getIfPresent(MemoryCacheService.CacheEnum.RESOURCE_CONDITIONAL_CREATE_VERSION, pid.getId());
+								if (version == null) {
+									version = myResourceTableDao.findCurrentVersionByPid(pid.getId());
+									if (version != null) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java:378]: Version can never be null here - This is the case for a conditional create where we discovered that the resource was already in the DB prior to our current transaction, which means it absolutely has to have an ID and a version. I'm going to add a whole bunch more comments here, I realize this block is kinda hard to follow.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java [hunk 3]
```diff
@@ -791,6 +811,10 @@ public void beforeCommit(boolean readOnly) {
 
 		ourLog.debug("Processed delete on {} (matched {} resource(s)) in {}ms", theUrl, deletedResources.size(), w.getMillis());
 
+		for (P next : theResourceIds) {
+			theTransactionDetails.addDeletedResourceId(next);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java:815]: good call, done

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java [hunk 4]
```diff
@@ -305,34 +303,20 @@ public class ResourceTable extends BaseHasResource implements Serializable, IBas
...
-			ResourceTable that = (ResourceTable) owner;
-			return that.myFhirId != null ? that.myFhirId : that.myId.toString();
-		}
-	}
-
 	@Version
 	@Column(name = "RES_VER")
 	private long myVersion;
-
 	@OneToMany(mappedBy = "myResourceTable", fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java:309]: Heh - I was trying to make it consistent with its neighbours, but looking at this class we seem to have a mix of blank lines and no-blank-lines. I agree a blank line is better, I've added one between all fields with annotations.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java [hunk 5]
```diff
@@ -343,6 +327,27 @@ public ResourceTable() {
 		super();
 	}
 
+	/**
+	 * If we've already updated the resource within the current transaction, we don't
+	 * want to update it a second time or create a second {@link ResourceHistoryTable}
+	 * entity for this resource. This could happen if a resource was deleted and
+	 * resurrected within a single FHIR transaction bundle, which is legal to do.
+	 * We can't actually create multiple versions of a resource within a single
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java:335]: Reworked

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/storage/TransactionDetails.java [hunk 6]
```diff
@@ -62,6 +56,7 @@ public class TransactionDetails {
 	private Map<String, IResourcePersistentId> myResolvedResourceIds = Collections.emptyMap();
 	private Map<String, IResourcePersistentId> myResolvedMatchUrls = Collections.emptyMap();
 	private Map<String, Supplier<IBaseResource>> myResolvedResources = Collections.emptyMap();
+	private Set<IResourcePersistentId> myDeletedResourceIds = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/storage/TransactionDetails.java:59]: The idea is just that most of these hashmaps will not get any entries in a typical flow (e.g. any given particular operation might add to one of them but not all of them) and a `new HashMap()` has a much bigger memory impact than `Collections.emptyMap()` so we might as well not create the mutable structure unless it's actually going to need to hold something.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java [hunk 7]
```diff
@@ -617,12 +627,15 @@ public DaoMethodOutcome delete(IIdType theId,
 			throw new ResourceVersionConflictException(Msg.code(961) + "Trying to delete " + theId + " but this is not the current version");
 		}
 
+		JpaPid persistentId = JpaPid.fromId(entity.getResourceId());
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java:630]: persistentId gets used on the immediate line below, which applies both to the condition below but also outside of it

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java]: Fair enough. I am curious about what use case requires a version increment when no properties have changed?
